src/codegen/mod.rs

```diff
@@ -14,30 +14,39 @@ use std::sync::atomic::{AtomicI32, Ordering};
 // therefore every subexpression stores a type
 pub fn gen(compil_unit: &desc::CompilUnit, idx_checks: bool) -> String {
     let fun_defs_to_be_generated = compil_unit
-        .fun_defs
+        .item_defs
         .iter()
-        .filter(|f| {
-            f.param_decls.iter().all(|p| {
-                !is_shape_ty(p.ty.as_ref().unwrap())
-                    && !matches!(
-                        &p.ty.as_ref().unwrap().ty,
-                        desc::TyKind::Data(desc::DataTy {
-                            dty: desc::DataTyKind::ThreadHierchy(_),
-                            ..
-                        })
-                    )
-            })
+        .filter_map(|item| {
+            match item {
+                Item::FunDef(f) =>
+                    if f.param_decls.iter().all(|p| {
+                        !is_shape_ty(p.ty.as_ref().unwrap())
+                            && !matches!(
+                                &p.ty.as_ref().unwrap().ty,
+                                desc::TyKind::Data(desc::DataTy {
+                                    dty: desc::DataTyKind::ThreadHierchy(_),
+                                    ..
+                                })
+                            )
+                        }) {
+                        Some(f)
```

Comment:
Was für Funktionen filter wie hier raus? Und wieso filtern wir die heraus?

Reply:
Functions that take thread hierarchies or views as arguments.
Roughly speeking, there is no way to straight forwardly generate code for these functions. The reason is that for views the accesses need to be inlined and the underlying array pointers must be passed as arguments; for thread hierarchies the parameters are completely removed during codegen.
That means, that a new function must be generated for every call of this top-level definition and the top-level definition itself should not be generated.

This should not be relevant for you (at least for now).